File: aoc-go/2021/day_07.go
package solutions

import (
	"fmt"
	"io/ioutil"
	"math"
	"strconv"
	"strings"
)

func disatance(positions []int64, position int64, accumulate bool) int64 {
	total := int64(0)
	for _, p := range positions {
		d := int64(math.Abs(float64(p - position)))
		if accumulate {
			total += d * (d + 1) / 2
		} else {
			total += d
		}
	}
	return total
}

func ShowD07() {
	data, _ := ioutil.ReadFile("../../input/day_07.in")
	input := strings.TrimSpace(string(data))
	crabPositions := []int64{}
	for _, n := range strings.Split(input, ",") {
		parse, _ := strconv.Atoi(n)
		crabPositions = append(crabPositions, int64(parse))
	}

	min, max := int64(10000), int64(0)
	for _, p := range crabPositions {
		if p < min {
			min = p
		}
		if p > max {
			max = p
		}
	}

	minFuel := int64(math.MaxInt64)
	for i := min; i < max; i++ {
		current := disatance(crabPositions, i, false)
		if current < minFuel {
			minFuel = current
		}
	}
	fmt.Printf("- Part 1: %d\n", minFuel)

	minFuelAcc := int64(math.MaxInt64)
	for i := min; i < max; i++ {
		current := disatance(crabPositions, i, true)
		if current < minFuelAcc {
			minFuelAcc = current
		}
	}

	fmt.Printf("- Part 2: %d\n", minFuelAcc)
}

File: aoc-rust/src/2021/day_07.rs
fn distance(positions: &Vec<u64>, position: u64, accumulate: bool) -> i64 {
    let mut distance_sum = 0;
    positions.iter().for_each(|p| {
        let d = (*p as i64 - position as i64).abs();
        if accumulate {
            distance_sum += d * (d + 1) / 2;
        } else {
            distance_sum += d;
        }
    });
    distance_sum
}

pub fn show() {
    let crabs_positions: Vec<u64> = include_str!("../../../input/day_07.in")
        .trim()
        .split(',')
        .map(|c| c.parse::<u64>().unwrap())
        .collect();

    let (min, max) = crabs_positions.iter().fold((1000, 0), |mut bounds, p| {
        if *p < bounds.0 {
            bounds.0 = *p
        }
        if *p > bounds.1 {
            bounds.1 = *p
        }
        bounds
    });

    let count = (min..max)
        .map(|x| distance(&crabs_positions, x, false))
        .min()
        .unwrap();
    println!("- Part 1: {}", count);

    let count = (min..max)
        .map(|x| distance(&crabs_positions, x, true))
        .min()
        .unwrap();
    println!("- Part 2: {}", count);
}
